src/meta/expr.lean

```diff
@@ -580,6 +580,11 @@ meta def to_implicit_binder : expr → expr
 | (pi n _ d b) := pi n binder_info.implicit d b
 | e  := e
 
+/--  Takes an `expr` and returns a list of its summands. -/
+meta def list_summands : expr → list expr
+| `(has_add.add %%a %%b) := a.list_summands ++ b.list_summands
+| a                      := [a]
```

Comment:
It's a lot uglier, but here's a more general implementation that works for other operations too:
```lean
meta def list_binary_operands_aux (f : expr) : expr → tactic (list expr)
| x@(expr.app (expr.app g a) b) := do
  some _ ← try_core (unify f g) | pure [x],
  as ← a.list_binary_operands_aux,
  bs ← a.list_binary_operands_aux,
  pure (as ++ bs)
| a                      := pure [a]

/-- Produces a list of all the operands in `x`, ignoring associativity. -/
meta def list_binary_operands (f : pexpr) (x : expr) : tactic (list expr) :=
do
  t ← infer_type x,
  fe ← to_expr ``(%%f : %%t → %%t → %%t ),
  list_binary_operands_aux fe x
```
`list_summands` is then just ```list_binary_operands ``((+))```

Reply:
Oh, that's funny: I was going to try this out!

```lean
meta def list_explicit_args (f : expr) : tactic (list expr) :=
tactic.fold_explicit_args f [] (λ ll e, return $ ll ++ [e])

--  this function appears in `move_add`
meta def list_named_fun (op : name) : expr → tactic (list expr) | f :=
if f.get_app_fn.const_name = op then do
  fargs ← f.list_explicit_args,
  fargs_rec ← fargs.mmap list_named_fun,
  return fargs_rec.join else
return [f]
```

Reply:
Your version works with binary operations.  I only tested mine with binary operations, but it might work more generally.

The other difference is that you provide a `pexpr` and I provide a `name`.  I am not really sure what is the difference betweem `name` and `pexpr`, though

Reply:
Matching a Pexpr is generally a better idea than a name, because it means you can match against an expression like `(add_powers 2) a b` which adds the squares of `a` and `b`

Reply:
With my version of the code, I get:
```
def ohmy (a b c : ℕ) : ℕ := sorry

example : ohmy (ohmy 3 4 5) 2 (ohmy 7 (ohmy 10 8 12) 9) = 3 * 13 :=
by do
  `(%%lhs = %%rhs) ← target,
  list_named_fun ``expr.ohmy lhs >>= trace,
  list_named_fun ``has_mul.mul rhs >>= trace
/-
[3, 4, 5, 2, 7, 10, 8, 12, 9]
[3, 13]
-/
```

Reply:
> . I only tested mine with binary operations, but it might work more generally.

I find it hard to beleive that this operation is at all useful for anything but binary functions!

Reply:
I think either of these implementations is fine. (Note, `f.get_app_fn.const_name = op` is the same as `f.is_app_of op`.) It bothers me slightly (aesthetically, really) that this function doesn't need to be tactic valued, but, whatever. It's a lot easier to implement.

For Eric's version, I might make the `aux` version the primary one. If the user has the function as a `pexpr` they can elaborate it themselves.